src/rmt.rs

```diff
@@ -0,0 +1,415 @@
+use std::mem::ManuallyDrop;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Channel {
+    // TODO: Work out the different number of channels per chip.
+    Channel0,
+    Channel1,
+    Channel2,
+    Channel3,
+}
+
+impl Into<u32> for Channel {
+    fn into(self) -> u32 {
+        match self {
+            Channel::Channel0 => rmt_channel_t_RMT_CHANNEL_0,
+            Channel::Channel1 => rmt_channel_t_RMT_CHANNEL_1,
+            Channel::Channel2 => rmt_channel_t_RMT_CHANNEL_2,
+            Channel::Channel3 => rmt_channel_t_RMT_CHANNEL_3,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct Pulse {
+    ticks: PulseTicks,
+    pin_state: PinState,
+}
+
+impl Pulse {
+    pub fn new(pin_state: PinState, ticks: PulseTicks) -> Self {
+        Pulse { pin_state, ticks }
+    }
+
+    /// Create a `Pulse` using a `Duration`.
+    ///
+    /// We need to know the ticks frequency (`ticks_hz`), which depends on the prepared channel
+    /// within a `Writer`. To get the frequency for the `ticks_hz` argument, use
+    /// `Writer::counter_clock()`. For example:
+    /// ```
+    /// # use esp_idf_sys::EspError;
+    /// # use esp_idf_hal::gpio::Output;
+    /// # use esp_idf_hal::rmt::Channel::Channel0;
+    /// # fn example() -> Result<(), EspError> {
+    /// # let peripherals = Peripherals::take()?;
+    /// # let led: Gpio18<Output> = peripherals.pins.gpio18.into_output()?;
+    /// let mut writer = Writer::new(led, cfg)?;
+    /// let ticks_hz = writer.counter_clock()?;
+    /// let pulse = Pulse::new_with_duration(ticks_hz, PinState::High, Duration::from_nanos(500))?;
+    /// # }
+    /// ```
+    pub fn new_with_duration(
+        ticks_hz: u32,
+        pin_state: PinState,
+        duration: Duration,
+    ) -> Result<Self, EspError> {
+        let ticks = PulseTicks::new_with_duration(ticks_hz, duration)?;
+        Ok(Self::new(pin_state, ticks))
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct PulseTicks(u16);
+
+impl PulseTicks {
+    const MAX: u16 = 32767;
+
+    /// PulseTick needs to be unsigned 15 bits: 0-32767 inclusive.
+    pub fn new(v: u16) -> Result<Self, EspError> {
+        if v > Self::MAX {
+            Err(EspError::from(ESP_ERR_INVALID_ARG as i32).unwrap())
+        } else {
+            Ok(Self(v))
+        }
+    }
+
+    /// Use the maximum value of 32767.
+    pub fn max() -> Self {
+        Self(Self::MAX)
+    }
+
+    /// Convert a `Duration` into `PulseTicks`.
+    ///
+    pub fn new_with_duration(ticks_hz: u32, duration: Duration) -> Result<Self, EspError> {
+        let ticks = duration
+            .as_nanos()
+            .checked_mul(ticks_hz as u128)
+            .ok_or(EspError::from(EOVERFLOW as i32).unwrap())?
+            / 1_000_000_000;
+        let ticks = u16::try_from(ticks).map_err(|_| EspError::from(EOVERFLOW as i32).unwrap())?;
+
+        Self::new(ticks)
+    }
+}
+
+pub mod config {
+    use embedded_hal::digital::PinState;
```

Comment:
Nit, but do you absolutely need to use this? The problem is, we support *both* embedded HALs now (V0.2 and V1.0). By using `embedded_hal::digital::PinState` you are basically asking users to import and use a trait from e-hal V1.0., while they might be using the traits from e-hal 0.2 for everything else. We can use a bool here (with proper renaming of e.g. `carrier_level` to `carrier_level_high` or a simple enum of ours.

Reply:
A crate enum is good. It's what I used initially.

I was just using `embedded_hal` as the "ideal" set of types to use. I didn't realise it would cause dependency hell issues.